### ISSUE TITLE
[WSC] trying to put xcat common openbmc attributes into group def fails when createing node def

### ISSUE THREAD

**Issue by agibm** [xcat2/xcat-core#4039]
we have the following group definition:
````
[root@c699mgt00 create_cnodes]# lsdef -t group ws_compute -z
# <xCAT data object stanza file>

ws_compute:
    objtype=group
    cons=openbmc
    consoleondemand=1
    grouptype=static
    mgt=openbmc
    netboot=petitboot
    nichostnamesuffixes.ib2=-ib2
    nichostnamesuffixes.ib0=-ib0
    nichostnamesuffixes.ib1=-ib1
    nichostnamesuffixes.ib3=-ib3
    nicnetworks.ib2=IB02
    nicnetworks.ib0=IB00
    nicnetworks.ib1=IB01
    nicnetworks.ib3=IB03
    nictypes.ib2=Infiniband
    nictypes.ib0=Infiniband
    nictypes.ib1=Infiniband
    nictypes.ib3=Infiniband
    primarynic=mac
    profile=compute
    usercomment=[offline] -- from ws_compute group def
````
We then try to do:
````
chdef -z <<EOF
c699c001:
    objtype=node
    nodetype=mp
    bmc=10.3.1.101
    groups=all,compute,lsf,rh74_compute,witherspoon,ws_compute,ws_sequoiac699f01
    hostnames=c699f01-U04
    ip=10.3.1.1
    nicips.ib0=10.10.1.1
    nicips.ib1=10.11.1.1
    nicips.ib2=10.12.1.1
    nicips.ib3=10.13.1.1
    rack=c699f01
    switch=c699tor01
    switchport=1
    unit=u04
EOF
````
it then tells us:

Cannot set the 'bmc' attribute unless a value is provided for 'mgt'.
mgt => The method to use to do general hardware management of the node.  This attribute is used as the default if power or getmac is not set.  Valid values: openbmc, ipmi, blade, hmc, ivm, fsp, bpa, kvm, esx, rhevm.  See the power attribute for more details.
Error: Cannot set the attr='bmc' attribute unless 'mgt=openbmc'.
1 object definitions have been created or modified.
New object definitions 'c699c001' have been created.

The resultant definition is:
````
[root@c699mgt00 create_cnodes]# lsdef c699c001
Object name: c699c001
    cons=openbmc
    consoleondemand=1
    groups=all,compute,lsf,rh74_compute,witherspoon,ws_compute,ws_sequoiac699f01
    hostnames=c699f01-U04
    ip=10.3.1.1
    mgt=openbmc
    netboot=petitboot
    nichostnamesuffixes.ib2=-ib2
    nichostnamesuffixes.ib0=-ib0
    nichostnamesuffixes.ib1=-ib1
    nichostnamesuffixes.ib3=-ib3
    nicips.ib2=10.12.1.1
    nicips.ib0=10.10.1.1
    nicips.ib1=10.11.1.1
    nicips.ib3=10.13.1.1
    nicnetworks.ib2=IB02
    nicnetworks.ib0=IB00
    nicnetworks.ib1=IB01
    nicnetworks.ib3=IB03
    nictypes.ib2=Infiniband
    nictypes.ib0=Infiniband
    nictypes.ib1=Infiniband
    nictypes.ib3=Infiniband
    nodetype=mp
    postbootscripts=otherpkgs
    primarynic=mac
    profile=compute
    rack=c699f01
    switch=c699tor01
    switchport=1
    unit=u04
````
note, If i repeat the command a second time it works:
````
[root@c699mgt00 create_cnodes]# chdef -z <<EOF
c699c001:
    objtype=node
    nodetype=mp
    bmc=10.3.1.101
    groups=all,compute,lsf,rh74_compute,witherspoon,ws_compute,ws_sequoiac699f01
    hostnames=c699f01-U04
    ip=10.3.1.1
    nicips.ib0=10.10.1.1
    nicips.ib1=10.11.1.1
    nicips.ib2=10.12.1.1
    nicips.ib3=10.13.1.1
    rack=c699f01
    switch=c699tor01
    switchport=1
    unit=u04
EOF
1 object definitions have been created or modified.
````

**Comment by bybai** [comment]
Hi @agibm,
I want to confirm :  although there are mgt error messages,  executing `chdef -z` at first time can create the correct node definition, is it right?
Since the following `chdef -z` can work well, it is not a block issue, it is in the backlog list and will look into it later. Thanks.

**Comment by whowutwut** [comment]
@bybai The issue here is that running the same command, the first time it failed with ERRORs and the 2nd time it succeeds.   While it does not block , it's inconsistent and not good for usability.

**Comment by bybai** [comment]
@whowutwut , will debug it next day. Thanks.

**Comment by bybai** [comment]
Hi @agibm, this issue is fixed in $4185, thanks.

**Comment by robin2008** [comment]
@bybai I thought the error message you provided might not be valid and it will cause confusing.

> [root@bybc0605 tools]# /opt/xcat/bin/chdef c699c001 bmc=10.3.1.101 groups=all
Error: Cannot set the attr='bmc' attribute unless 'mgt=openbmc'.

For example, the node could also be set with `bmc` attribute with `mgt=ipmi`

**Comment by bybai** [comment]
@robin2008, The error message is from old code,  is this description better "Cannot set the 'bmc' attribute unless a value is provided for mgt attribute as 'mgt=ipmi'."
BTW, 'mgt=openbmc' or 'mgt=ipmi' is calculated by code logic.

**Comment by robin2008** [comment]
Obviously the previous code logic has bug and it give an inaccurate error message to user.
Here I define an x64 ipmi node, and the complain is still `mgt=openbmc`
```
mkdef fake1 ip=1.1.1.1 arch=x86_64 netboot=xnba groups=all
chdef fake1 bmc=10.3.1.101 groups=all
```

**Comment by bybai** [comment]
@robin2008, the 'mgt=openbmc' is calculated by the code, in this issue, I fixed duplicate error message and added the consideration of groups attributes checks. I noticed there were many lines codes to get the condition 'mgt=<value>', in this issue, I regarded the way to get condition as correct by default, so it was another issue you found now. Generally speaking, I think it is hard to check mgt value. For example, `chdef fake1 bmc=1.1.1.1 groups=all`, if there is no 'arch', we cannot give the recommended value for 'mgt'.  In a word, let me open a new issue to trace the algorithm how to calculate attribute preposition attribute correct value in 'chdef'.
